imap_processing/spice/geometry.py

```diff
@@ -149,13 +149,45 @@ def get_spacecraft_spin_phase(
     spin_phase : float or np.ndarray
         Spin phase for the input query times.
     """
+    spin_df = get_spin_data()
+
+    # Combine spin_start_sec and spin_start_subsec to get the spin start
+    # time in seconds. The spin start subseconds are in milliseconds.
+    # TODO: Decide if we should do this calculation in the data itself
+    # or in get_spin_data function.
+    spin_df["spin_start_time"] = (
+        spin_df["spin_start_sec"] + spin_df["spin_start_subsec"] / 1e3
+    )
+
     if isinstance(query_met_times, float):
-        # TODO: call get_spin_data function to get spin data for the
-        # input query times
-        # Here, return a float (dummy implementation)
-        return 0.0  # Replace this with actual logic to calculate spin phase
-
-    # Return an ndarray of the same shape, filled with 0.0 for
-    # now (dummy implementation)
-    # TODO: Replace with actual logic to calculate spin phase
-    return np.array(query_met_times, dtype=float)
+        # calculate spin phase for a single query time
+        mask = spin_df["spin_start_time"] <= query_met_times
+        if mask.any():
+            last_spin = spin_df[mask].iloc[-1]
```

Comment:
As we discussed in the ENA tagup, we need a check to make sure this last_spin index doesn't index a previous spin due to a missing spin in the spin table. It also occurs to me that we may want/need to filter out bad spins using the `valid` flags present in the table. Since we are still gathering information about what those flags mean, perhaps this should just be captured in a TODO and github ticket.

Reply:
About your second point about filtering bad spin, I was just thinking about this. I was thinking of returning pd.DataFrame with spin information and spin phase calculated for the input times and let user decide what they want to do with spin info or filter spin phase based on bad spin or just use spin phase column from returned pd.DataFrame. What do you guys think?

Reply:
I would vote for just returning the spin phase for now. I think that later we might want to add in returning a quality flag along with each of the spin phase values. I think that returning too much will promote duplicated code each place this function is called rather than keeping the code common.

Reply:
created this issue for future work: https://github.com/IMAP-Science-Operations-Center/imap_processing/issues/863